File: Rotate Array.java
class Solution {
    public void rotate(int[] a, int k) {
        int n=a.length;
        k=k%n;
        int i=0;
        int j=n-1;
        //reverse whole array
        while(i<j){
            int temp=a[j];
            a[j]=a[i];
            a[i]=temp;
            i++;
            j--;
        }
        
        //reverse upto k
        i=0;
        j=k-1;
        while(i<j){
            int temp=a[j];
            a[j]=a[i];
            a[i]=temp;
            i++;
            j--;
        }
        
        //reverse next kth part
        i=k;
        j=n-1;
        while(i<j){
            int temp=a[j];
            a[j]=a[i];
            a[i]=temp;
            i++;
            j--;
        }
    }
}

File: rotate-array/rotate-array.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {        
        // eg 1 2 3 4 5 6 7
        k %= nums.size();
        reverse(nums.begin(), nums.end()); // reverse full 7 6 5 4 3 2 1
        reverse(nums.begin(), nums.begin()+k); // reverse till first k 5 6 7 4 3 2 1
        reverse(nums.begin()+k, nums.end()); // reverse till end from k 5 6 7 1 2 3 4        
    }
};
